### ISSUE TITLE
[Bug]: set state in Selection utility will make components (i.e. DetailsList) unselectable

### ISSUE THREAD

**Issue by AkiraVoid** [microsoft/fluentui#23271]
### Library

React / v8 (@fluentui/react)

### System Info

```shell
System:
    OS: Windows 10 10.0.22000
    CPU: (8) x64 Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
    Memory: 4.25 GB / 15.88 GB
  Browsers:
    Edge: Spartan (44.22000.120.0), Chromium (101.0.1210.53)
    Internet Explorer: 11.0.22000.120
  npmPackages:
    @fluentui/react: 8.64.1 => 8.64.1
    @fluentui/react-hooks: ^8.5.5 => 8.5.5
    @fluentui/react-icons-mdl2: 1.3.4 => 1.3.4
```

### Are you reporting Accessibility issue?

no

### Reproduction

https://codesandbox.io/s/fluentui-bug-set-state-in-selection-dykn7x

## Actual Behavior

When using `Selection` utility in a functional component (not test in class component) and set a state when `onSelectionChanged` fired, all components related to this `Selection` object (they are `MarqueeSelection` and `DetailsList` in repro above) will be unable to select.

## Expected Behavior

It should be available to set states in `onSelectionChanged` event. We have this requirement and I suppose everyone will have.

### Logs

```shell
There're no logs.
```

### Requested priority

High

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

**Comment by AkiraVoid** [comment]
I use this event to get selected items and store them into a state, for now I have a workaround: 

```
const selection = ...;

// Instead of store them into state, just use:
selection.getSelection();
// everywhere I want to use selected items.
```

**Comment by AkiraVoid** [comment]
Please I really need this to be fixed, or is there someone can help me with a workaround which could let me do some real-time detection on selecting? In the example from docs everything goes ok so I guess it will only happen in functional components.

@ecraig12345 @micahgodbolt

**Comment by AkiraVoid** [comment]
I found this will only happen without `setKey` property or with variable `setKey` property. The doc of `setKey` said if this property were changed, the selection will be reset, so I know why we should not change `setKey` every time we selected some items, but I don't know why we must provide this value, and it haven't been mentioned in docs.

**Comment by gouttierre** [comment]
@AkiraVoid - Thanks for filing this issue and updating its current status. Happy to see you managed to unblocked yourself with a with a workaround (https://github.com/microsoft/fluentui/issues/23271#issuecomment-1138210723). 
I’ll mark this as a feature request and add it to our backlog, but we’re likely not going to get to it for a while as we’re currently focused on larger coherence work items and enhancements to our performance. We will review the issues and update the control's docs.

**Comment by gar1t** [comment]
The [Selection example](https://developer.microsoft.com/en-us/fluentui#/controls/web/selection) shows the selection API within a function component - and this can be made to work with React state updates (see below). The [other examples I've seen](https://developer.microsoft.com/en-us/fluentui#/controls/web/detailslist), where state updates work fine, are all class based components.

To test state updates, modify `onSelectionChanged` in the `Selection` constructor to set state. E.g.

``` typescript
  const [count, setCount] = React.useState(0);

  const selection = React.useMemo(
    () =>
      new Selection({
        onSelectionChanged: () => {
           setCount(selection.count);
        },
        items,
      }),
    [items],
  );

  // In the compoment JSX...

  <p>{count}</p>
```

In my own example, which uses `Selection` to handle selects in a `DetailsList`, I observed the same behavior described in the OP. Unless FluentUI officially doesn't support React function components, I would consider this a bug and not a feature request. The ability to handle selection changes by updating application state is a mainstream use case.

To get my example to work, I needed to use `useConst` from `@fluentui/react-hooks` to define my list of items. As this is just an example, that approach works fine. I expect to use `useMemo` for real data, which will not be hard-coded. I haven't confirmed the behavior with `useMemo` however.

As an aside, I would expect `onSelectionChanged` to be called with the selection object as an argument - perhaps along with event information - as per other event related callbacks. There may be an obscure/surprising reason it's not, given the odd behavior here - but it ought to be IMO. I've never seen an event-handler callback that wasn't passed information about what happened.
